statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+        scope: stats::Scope,
+        client: T,
+        endpoint: &str,
+        config: PromBackendClientConfig,
+    ) -> Self {
+        let client = Arc::new(client);
+        let stats = Stats::new(scope);
+        let (sender, recv) = mpsc::channel::<Pdu>(config.max_queue);
+        // Currently, we need this tripwire to abort request retries. This can probably be refactored
+        let (trig, trip) = Tripwire::new();
+        let joinhandle = tokio::spawn(client_task(stats, client, recv, trip, config));
+
+        let inner = Arc::new(Mutex::new(PromClientInner {
+            joinhandle: Some(joinhandle),
+            _trig: trig,
+        }));
+
+        Self {
+            endpoint: String::from(endpoint),
+            inner,
+            sender,
+        }
+    }
+}
+
+impl BackendClient for PromClient {
+    fn sender(&self) -> mpsc::Sender<Pdu> {
+        self.sender.clone()
+    }
+
+    fn endpoint(&self) -> &str {
+        self.endpoint.as_str()
+    }
+
+    // Request the join handle for this client, which completes when all
+    // operations attempting to use this client are complete and client has shut
+    // down. Note that this is a destructive action, and a second call to a join
+    // handle will return None.
+    fn join_handle(&self) -> Option<JoinHandle<()>> {
+        let mut lock = self.inner.lock().unwrap();
+        lock.join_handle()
+    }
+}
+
+struct PromClientInner {
+    joinhandle: Option<JoinHandle<()>>,
+    // Once this trigger is dropped, request retries will be cancelled.
+    _trig: Trigger,
+}
+
+impl PromClientInner {
+    // Destructively move the join handle out of the client
+    pub fn join_handle(&mut self) -> Option<JoinHandle<()>> {
+        self.joinhandle.take()
+    }
+}
+
+/// A thin client wrapper used for mocking in tests
+#[async_trait]
+trait PromWriteRequestClient {
+    async fn send_write_request(
+        &self,
+        compressed_write_request: Vec<u8>,
+    ) -> Result<reqwest::Response, reqwest::Error>;
+}
+
+#[derive(Clone, Debug)]
+struct ReqwestPromWriteRequestClient {
+    inner: Client,
+    endpoint: String,
+}
+
+#[async_trait]
+impl PromWriteRequestClient for ReqwestPromWriteRequestClient {
+    async fn send_write_request(
+        &self,
+        compressed_write_request: Vec<u8>,
+    ) -> Result<reqwest::Response, reqwest::Error> {
+        self.inner
+            .post(self.endpoint.clone())
+            .header("Content-Type", "application/x-protobuf")
+            .header("Content-Encoding", "snappy")
+            .header("X-Prometheus-Remote-Write-Version", "0.1.0")
+            .body(compressed_write_request)
+            .send()
+            .await
+    }
+}
+
+async fn client_task<T: PromWriteRequestClient + Send + Sync + 'static>(
+    stats: Stats,
+    client: Arc<T>,
+    mut rx: mpsc::Receiver<Pdu>,
+    trip: Tripwire,
+    config: PromBackendClientConfig,
+) {
+    let semaphore = Arc::new(Semaphore::new(config.max_in_flight));
```

Comment:
Using max_in_flight at this layer means that config field is defining the maximum number of requests in flight _per client task_, correct? How does this interact with the shard ring of clients in the Backend type?